runInLinuxVM: fix simple build

wolfgangwalther opened the pull request:
In 58570e75d9bddede2a854b0f7cb8df7f1818c541, I moved the PATH for coreutils *before* the sourcing of attrs + stdenv. However, AFAICT, this leads to PATH being overwritten and cat+chmod can't be found later on.

We don't need to set PATH here at all, we can just do the same thing as in stage2Init - reference every coreutil explicitly. This keeps the environment the cleanest.

Resolves #378545


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
runInLinuxVM: fix simple build

In 58570e75d9bddede2a854b0f7cb8df7f1818c541, I moved the PATH for
coreutils *before* the sourcing of attrs + stdenv. However, AFAICT, this
leads to PATH being overwritten and cat+chmod can't be found later on.

We don't need to set PATH here at all, we can just do the same thing as
in stage2Init - reference every coreutil explicitly. This keeps the
environment the cleanest.

Resolves #378545

malte-v reviewed (approved):
Can confirm that this fixes my simple test derivation and also the more involved use-case that I originally encountered the bug with. Thanks for resolving this so quickly! :)

Mic92 commented on pkgs/build-support/vm/default.nix:
```diff
@@ -258,7 +257,7 @@ rec {
     # debug inside the VM if the build fails (when Nix is called with
     # the -K option to preserve the temporary build directory).
-    cat > ./run-vm <<EOF
+    ${coreutils}/bin/cat > ./run-vm <<EOF
```
Not sure what splicing is in this file, but intuitivly I would say this should be:

```suggestion
    ${lib.getBin buildPackages.coreutils}/bin/cat > ./run-vm <<EOF
```

Mic92 replied:
But looks like this overall doesn't cross-compile well.

Mic92 commented on pkgs/build-support/vm/default.nix:
```diff
@@ -237,7 +237,6 @@ rec {
-    PATH=${coreutils}/bin
 
     if [ -f "''${NIX_ATTRS_SH_FILE-}" ]; then
       cp $NIX_ATTRS_JSON_FILE $NIX_ATTRS_SH_FILE xchg
```
What about this?

Mic92 commented on pkgs/build-support/vm/default.nix:
```diff
@@ -258,7 +257,7 @@ rec {
     # the -K option to preserve the temporary build directory).
-    cat > ./run-vm <<EOF
+    ${coreutils}/bin/cat > ./run-vm <<EOF
     #! ${bash}/bin/sh
```
Unrelated to this PR (doesn't need to be addressed):

```suggestion
    #! ${runtimeShell}
```

wolfgangwalther committed:
nixos/lib/make-*-disk-zfs-image: fix build

Those two helpers have not received the memo about a new requirement for
virtiofs and about changes regarding the zfs related attributes.

Can be tested with nix/release.nix' amazonImageZfs.

Ma27 commented:
> They multi-variant can be tested with nix-build nixos/release.nix -A amazonImageZfs. This must have been broken for longer, I'm unclear why, because it is part of release.nix? Doesn't that mean it should not be broken? Edit: Also see #378799, which is equally odd that it was unnoticed.

IIRC AMIs get built in another repository these days: https://github.com/nixos/amis

Now idea how this interacts with nixpkgs though.
cc @arianvp any opinions on the amazon images still being in release.nix?

arianvp commented:
We fetch the image from the hydra jobs at the moment so it needs to be in `release.nix`.   Me and @Mic92 are working on an initiative to move *all* images out of hydra but this is not being worked on actively yet.

The ZFS variant of the image is unmaintained and unused. We do not upload ZFS images. I don't care about ZFS and have no interest in fixing bugs in that image builder as it's kinda cursed. But people got loud when I suggested removing it.

arianvp commented:
I suggest removing `amazonImageZfs` at least. I am not maintaining it. And seems nobody else is either.   We need to keep `.amazonImage` for the time being though

Ma27 reviewed (approved):
LGTM.

And I'd be fine with removing the Amazon ZFS image.

Mic92 commented:
Do we still build any amazon images in hydra? We should disable them there.

arianvp commented:
> Do we still build any amazon images in hydra? We should disable them there.

Yes and currently the https://github.com/nixos/amis repo depends on that.  We can switch over now that Github Actions supports ARM though.
